File: iosApp/iosApp/Features/Onboarding/OnboardingContent.swift
//
//  OnboardingContent.swift
//  iosApp
//
//  Created by velkonost on 30.01.2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK
import Lottie

struct OnboardingContent: View {
    
    private let isLoading: Bool
    private let title: String
    private let step: Int
    private let onNextClick: () -> Void
    private let onSkipClick: () -> Void
    
    @State private var firstPointVisible: Bool = false
    @State private var secondPointVisible: Bool = false
    @State private var thirdPointVisible: Bool = false
    @State private var forthPointVisible: Bool = false
    @State private var fifthPointVisible: Bool = false
    
    @State private var buttonVisible: Bool = false
    
    init(isLoading: Bool, title: String, step: Int, onNextClick: @escaping () -> Void, onSkipClick: @escaping () -> Void) {
        self.isLoading = isLoading
        self.title = title
        self.step = step
        self.onNextClick = onNextClick
        self.onSkipClick = onSkipClick
    }
    
    var body: some View {
        VStack(spacing: 0) {
            Spacer().frame(height: 60)
            
            HStack {
                Spacer()
                Text(SharedR.strings().skip_btn.desc().localized())
                    .style(.bodySmall)
                    .foregroundColor(.textPrimary)
                    .padding(.trailing, 16)
                    .onTapGesture {
                        onSkipClick()
                    }
            }
            .padding(.bottom, 20)
            
            Spacer().frame(height: 24)
            
            OnboardingLottieView(resource: SharedR.files().anim_onboarding_1.path, loopMode: .loop)
                .frame(width: 256, height: 256)
                .scaleEffect(0.3)
            
            Spacer().frame(height: 24)
            
            if firstPointVisible {
                OnboardingPoint(
                    title: SharedR.strings().onboarding_step_1.desc().localized(),
                    index: 0
                )
            }
            
            if secondPointVisible {
                OnboardingPoint(
                    title: SharedR.strings().onboarding_step_2.desc().localized(),
                    index: 1
                )
            }
            
            if thirdPointVisible {
                OnboardingPoint(
                    title: SharedR.strings().onboarding_step_3.desc().localized(),
                    index: 2
                )
            }
            
            if forthPointVisible {
                OnboardingPoint(
                    title: SharedR.strings().onboarding_step_4.desc().localized(),
                    index: 3
                )
            }
            
            if fifthPointVisible {
                OnboardingPoint(
                    title: SharedR.strings().onboarding_step_5.desc().localized(),
                    index: 4
                )
            }
         
            Spacer()
            
            if buttonVisible {
                AppButton(
                    labelText: step == 5 ? SharedR.strings().onboarding_btn.desc().localized() : SharedR.strings().continue_btn.desc().localized(),
                    isLoading: isLoading,
                    onClick: onNextClick
                )
            }
            
            Spacer().frame(height: 48)
        }
        .onAppear {
            withAnimation(.bouncy.delay(1)) {
                firstPointVisible = true
            }
            
            withAnimation(.bouncy.delay(3)) {
                secondPointVisible = true
            }
            
            withAnimation(.bouncy.delay(5)) {
                thirdPointVisible = true
            }
            
            withAnimation(.bouncy.delay(7)) {
                forthPointVisible = true
            }
            
            withAnimation(.bouncy.delay(9)) {
                fifthPointVisible = true
            }
            
            withAnimation(.bouncy.delay(11)) {
                buttonVisible = true
            }
        }
    }
}

struct OnboardingPoint: View {
    
    private let index: Int
    private let title: String
    
    init(title: String, index: Int) {
        self.title = title
        self.index = index
    }
    
    @State private var animVisible: Bool = false
    
    var body: some View {
        HStack(spacing: 0) {
            if animVisible {
                OnboardingLottieView(resource: SharedR.files().anim_mark.path)
                    .frame(width: 24, height: 24)
                    .scaleEffect(0.1)
            } else {
                Spacer().frame(width: 24, height: 24)
            }
            
            Text(title.uppercased())
                .style(.labelMedium)
                .foregroundColor(.textTitle)
                .multilineTextAlignment(.leading)
                .padding(.leading, 8)
            
            Spacer()
                
        }
        .padding(.top, 12)
        .padding(.horizontal, 32)
        .onAppear {
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.5 + 2.0 * Double(index)) {
                withAnimation(.bouncy) {
                    animVisible = true
                }
            }
            
        }
    }
}

struct OnboardingLottieView: UIViewRepresentable {
    
    var resource: String

    var loopMode: LottieLoopMode = .playOnce
    
    func updateUIView(_ uiView: UIViewType, context: Context) {}

    func makeUIView(context: Context) -> Lottie.LottieAnimationView {
        let animationView = LottieAnimationView(filePath: resource)
        
        animationView.contentMode = .scaleAspectFit
        animationView.loopMode = loopMode
        animationView.play()
        
        return animationView
    }
    
}

File: androidApp/features/onboarding/src/main/kotlin/com/velkonost/getbetter/android/features/onboarding/OnboardingContent.kt
package com.velkonost.getbetter.android.features.onboarding

import androidx.compose.animation.AnimatedContent
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.airbnb.lottie.compose.LottieAnimation
import com.airbnb.lottie.compose.LottieCompositionSpec
import com.airbnb.lottie.compose.LottieConstants
import com.airbnb.lottie.compose.rememberLottieComposition
import com.velkonost.getbetter.core.compose.components.AppButton
import com.velkonost.getbetter.shared.resources.SharedR
import dev.icerock.moko.resources.compose.colorResource
import dev.icerock.moko.resources.compose.stringResource
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

@Composable
fun OnboardingContent(
    modifier: Modifier = Modifier,
    isLoading: Boolean,
    title: String,
    step: Int,
    onNextClick: () -> Unit,
    onSkipClick: () -> Unit
) {
    val composition by rememberLottieComposition(
        LottieCompositionSpec.RawRes(SharedR.files.anim_onboarding_1.rawResId)
    )

    val firstPointVisible = remember { mutableStateOf(false) }
    val secondPointVisible = remember { mutableStateOf(false) }
    val thirdPointVisible = remember { mutableStateOf(false) }
    val forthPointVisible = remember { mutableStateOf(false) }
    val fifthPointVisible = remember { mutableStateOf(false) }
    val buttonVisible = remember { mutableStateOf(false) }

    val scope = rememberCoroutineScope()

    LaunchedEffect(Unit) {
        scope.launch {
            delay(2000)
            firstPointVisible.value = true
            delay(2000)
            secondPointVisible.value = true
            delay(2000)
            thirdPointVisible.value = true
            delay(2000)
            forthPointVisible.value = true
            delay(2000)
            fifthPointVisible.value = true
        }
    }

    Column(
        modifier = modifier.fillMaxSize(),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {

        Spacer(modifier.height(50.dp))

        Row {
            Spacer(modifier.weight(1f))
            Text(
                modifier = modifier
                    .padding(end = 16.dp)
                    .clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                        onClick = onSkipClick
                    ),
                text = stringResource(resource = SharedR.strings.skip_btn),
                style = MaterialTheme.typography.bodySmall,
                color = colorResource(resource = SharedR.colors.text_primary)
            )
        }

        Spacer(modifier.height(24.dp))

        LottieAnimation(
            modifier = modifier.height(256.dp),
            composition = composition,
            iterations = LottieConstants.IterateForever,
        )

        Spacer(modifier.height(32.dp))

        AnimatedContent(targetState = firstPointVisible, label = "") {
            if (it.value) {
                OnboardingPoint(title = stringResource(resource = SharedR.strings.onboarding_step_1))
            }
        }

        AnimatedContent(targetState = secondPointVisible, label = "") {
            if (it.value) {
                OnboardingPoint(title = stringResource(resource = SharedR.strings.onboarding_step_2))
            }
        }

        AnimatedContent(targetState = thirdPointVisible, label = "") {
            if (it.value) {
                OnboardingPoint(title = stringResource(resource = SharedR.strings.onboarding_step_3))
            }
        }

        AnimatedContent(targetState = forthPointVisible, label = "") {
            if (it.value) {
                OnboardingPoint(title = stringResource(resource = SharedR.strings.onboarding_step_4))
            }
        }

        AnimatedContent(targetState = fifthPointVisible, label = "") {
            if (it.value) {
                OnboardingPoint(title = stringResource(resource = SharedR.strings.onboarding_step_5))
            }
        }

        Spacer(modifier.weight(1f))

        AnimatedContent(targetState = buttonVisible, label = "") {
            if (it.value) {
                AppButton(
                    modifier = modifier
                        .align(Alignment.CenterHorizontally),
                    labelText = stringResource(resource = SharedR.strings.onboarding_btn),
                    isLoading = isLoading,
                    onClick = onNextClick
                )
            }
        }

        Spacer(modifier = modifier.height(64.dp))

    }
}

@Composable
fun OnboardingPoint(
    modifier: Modifier = Modifier,
    title: String
) {
    val composition by rememberLottieComposition(
        LottieCompositionSpec.RawRes(SharedR.files.anim_mark.rawResId)
    )

    val scope = rememberCoroutineScope()
    val animVisible = remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        scope.launch {
            delay(1500)
            animVisible.value = true
        }
    }

    Row(
        modifier = modifier
            .padding(top = 16.dp, start = 16.dp, end = 16.dp)
    ) {
        if (animVisible.value) {
            LottieAnimation(
                modifier = modifier.size(24.dp),
                composition = composition,
                iterations = 1,
            )
        } else {
            Spacer(modifier.size(24.dp))
        }

        Text(
            modifier = modifier.padding(start = 8.dp),
            text = title.uppercase(),
            style = MaterialTheme.typography.labelMedium,
            color = colorResource(resource = SharedR.colors.text_title),
            textAlign = TextAlign.Start,
        )
    }

}